### PR TITLE
[TR] PIM-4261: introduce FileStorage component

### PR DESCRIPTION
**Pull request by jjanvier**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | LOL
| Tests pass?          | Y
| Scenarios pass?      | running
| Checkstyle issues?*  | cs-fixer launched
| PMD issues?**        | N
| Changelog updated?   | -
| Fixed tickets        | -
| DB schema updated?   | Y
| Migration script?    | to be done later
| Doc PR               |

### CODE REVIEW

File: src/Akeneo/Bundle/FileStorageBundle/DependencyInjection/AkeneoFileStorageExtension.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+/*
+ * This file is part of the Akeneo PIM Enterprise Edition.
+ *
+ * (c) 2014 Akeneo SAS (http://www.akeneo.com)
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
```
Comment (sumbobyboys) [inline src/Akeneo/Bundle/FileStorageBundle/DependencyInjection/AkeneoFileStorageExtension.php]: change licence :)

File: src/Akeneo/Bundle/FileStorageBundle/AkeneoFileStorageBundle.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the Akeneo PIM Enterprise Edition.
+ *
+ * (c) 2014 Akeneo SAS (http://www.akeneo.com)
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
```
Comment (sumbobyboys) [inline src/Akeneo/Bundle/FileStorageBundle/AkeneoFileStorageBundle.php]: licence :)

File: src/Akeneo/Bundle/FileStorageBundle/DependencyInjection/Compiler/ResolveDoctrineTargetModelPass.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+/**
+ * This file is part of the Akeneo PIM Enterprise Edition.
+ *
+ * (c) 2015 Akeneo SAS (http://www.akeneo.com)
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
```
Comment (sumbobyboys) [inline src/Akeneo/Bundle/FileStorageBundle/DependencyInjection/Compiler/ResolveDoctrineTargetModelPass.php]: you got the idea

File: src/Akeneo/Component/FileStorage/Model/FileInterface.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     * @return UploadedFile
+     */
+    public function getUploadedFile();
+
+    /**
+     * @param UploadedFile $uploadedFile
+     *
+     * @return FileInterface|null
+     */
+    public function setUploadedFile(UploadedFile $uploadedFile = null);
```
Comment (sumbobyboys) [inline src/Akeneo/Component/FileStorage/Model/FileInterface.php]: Seems weird :/

File: src/Akeneo/Component/FileStorage/PathGenerator.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * @author Julien Janvier <jjanvier@akeneo.com>
+ */
+class PathGenerator implements PathGeneratorInterface
+{
+    /**
+     * Generate all the path data of a file. If the original file name exceeds 100 characters, it is truncated.
+     * The file name is sanitized, and a unique ID is prepended.
+     *
+     * For example, a file called "this i#s the é file.txt'" will produce the following output:
```
Comment (sumbobyboys) [inline src/Akeneo/Component/FileStorage/PathGenerator.php]: é ?

File: src/Akeneo/Bundle/FileStorageBundle/Resources/config/model/doctrine/File.orm.yml [hunk 6]
```diff
@@ -0,0 +1,35 @@
+Akeneo\Component\FileStorage\Model\File:
```
Comment (nidup) [inline src/Akeneo/Bundle/FileStorageBundle/Resources/config/model/doctrine/File.orm.yml]: you need this mapping in CE?

File: src/Akeneo/Bundle/FileStorageBundle/Resources/config/repositories.yml [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    akeneo_file_storage.repository.file.class: Akeneo\Bundle\FileStorageBundle\Doctrine\ORM\Repository\FileRepository
+
+services:
+    akeneo_file_storage.repository.file:
+        class: %akeneo_file_storage.repository.file.class%
+        factory_service: doctrine.orm.entity_manager
+        factory_method: getRepository
+        arguments: [%akeneo_file_storage.model.file.class%]
+        tags:
+            - { name: 'pim_repository' }
```
Comment (nidup) [inline src/Akeneo/Bundle/FileStorageBundle/Resources/config/repositories.yml]: weird to have 'pim_repository' tag in an Akeneo Bundle ;)

File: src/Akeneo/Bundle/FileStorageBundle/Resources/config/savers.yml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+parameters:
+    akeneo_file_storage.saver.file.class: Akeneo\Bundle\StorageUtilsBundle\Doctrine\Common\Saver\BaseSaver
+
+services:
+    akeneo_file_storage.saver.file:
+        class: %akeneo_file_storage.saver.file.class%
+        arguments:
+            - '@doctrine.orm.entity_manager'
+            - '@pim_catalog.saver.base_options_resolver'
```
Comment (nidup) [inline src/Akeneo/Bundle/FileStorageBundle/Resources/config/savers.yml]: we should maybe add the @pim_catalog.saver.base_options_resolver service in Akeneo StorageUtils Bundle

File: src/Akeneo/Component/FileStorage/Model/File.php [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    /** @var int */
+    protected $size;
+
+    /** @var string */
+    protected $extension;
+
+    /** @var string */
+    protected $storage;
+
+    //TODDO: check if we really need it
```
Comment (nidup) [inline src/Akeneo/Component/FileStorage/Model/File.php]: yep check and drop the comment if useless ;)

Comment (nidup) [inline src/Akeneo/Component/FileStorage/Model/File.php]: yep check and drop the comment if useless ;)

File: src/Akeneo/Component/FileStorage/RawFile/RawFileFetcher.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Akeneo\Component\FileStorage\RawFile;
+
+use Akeneo\Component\FileStorage\Model\FileInterface;
+use League\Flysystem\FilesystemInterface;
+use PimEnterprise\Component\ProductAsset\Exception\FileTransferException;
```
Comment (nidup) [inline src/Akeneo/Component/FileStorage/RawFile/RawFileFetcher.php]: PimEnterprise ?

File: src/Akeneo/Component/FileStorage/RawFile/RawFileFetcherInterface.php [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Akeneo\Component\FileStorage\RawFile;
+
+use Akeneo\Component\FileStorage\Model\FileInterface;
+use League\Flysystem\FilesystemInterface;
+use PimEnterprise\Component\ProductAsset\Exception\FileTransferException;
```
Comment (nidup) [inline src/Akeneo/Component/FileStorage/RawFile/RawFileFetcherInterface.php]: same

File: src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+ */
+
+namespace Akeneo\Component\FileStorage\RawFile;
+
+use Akeneo\Component\FileStorage\FileFactoryInterface;
+use Akeneo\Component\FileStorage\PathGeneratorInterface;
+use Akeneo\Component\StorageUtils\Saver\SaverInterface;
+use League\Flysystem\FileExistsException;
+use League\Flysystem\MountManager;
+use PimEnterprise\Component\ProductAsset\Exception\FileRemovalException;
```
Comment (nidup) [inline src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php]: PimEnterprise

File: src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+use PimEnterprise\Component\ProductAsset\Exception\FileTransferException;
```
Comment (nidup) [inline src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php]: PimEnterprise

**Comment by nidup** [comment]
GTM once comments fixed

**Comment by grena** [comment]
Please use `/*` for head phpdoc

File: src/Akeneo/Component/FileStorage/FileFactoryInterface.php [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+ * File factory, create a \Akeneo\Component\FileStorage\Model\FileInterface
+ *
+ * @author Julien Janvier <jjanvier@akeneo.com>
+ */
+interface FileFactoryInterface
+{
+    /**
+     * @param \SplFileInfo $rawFile
+     * @param array        $pathInfo
+     * @param string       $destFsAlias
```
Comment (jmleroux) [inline src/Akeneo/Component/FileStorage/FileFactoryInterface.php]: Could be more precise. What is a destFsAlias ?

File: src/Akeneo/Component/FileStorage/Model/FileInterface.php [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+{
+    /**
+     * @return int
+     */
+    public function getId();
+
+    /**
+     * @return string
+     */
+    public function getKey();
```
Comment (grena) [inline src/Akeneo/Component/FileStorage/Model/FileInterface.php]: Maybe add a small comment to explain what the key is, as it's not really obvious :)

File: src/Akeneo/Component/FileStorage/PathGenerator.php [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+     *
+     * @param \SplFileInfo $file
+     *
+     * @return array
+     */
+    public function generate(\SplFileInfo $file)
+    {
+        $originalFileName = ($file instanceof UploadedFile) ? $file->getClientOriginalName() : $file->getFilename();
+        $guid             = $this->generateId($originalFileName);
+        $sanitized        = preg_replace('#[^A-Za-z0-9\.]#', '_', $originalFileName);
```
Comment (jmleroux) [inline src/Akeneo/Component/FileStorage/PathGenerator.php]: you don't need to escape `.` in `[ ]`

**Comment by jmleroux** [comment]
A validator is maybe missing for FileType : 
```Akeneo\Component\FileStorage\Model\File:
    properties:
        uploadedFile:
            - File: ~
```

File: src/Akeneo/Bundle/FileStorageBundle/Resources/config/repositories.yml (hunk 7 above)
Comment (jjanvier) [inline src/Akeneo/Bundle/FileStorageBundle/Resources/config/repositories.yml]: nice catch

File: src/Akeneo/Component/FileStorage/Model/File.php (hunk 9 above)
Comment (jjanvier) [inline src/Akeneo/Component/FileStorage/Model/File.php]: right now, it's used for assets, but maybe we'll be able to drop it in a few weeks

File: src/Akeneo/Bundle/FileStorageBundle/Resources/config/savers.yml (hunk 8 above)
Comment (jjanvier) [inline src/Akeneo/Bundle/FileStorageBundle/Resources/config/savers.yml]: done

File: src/Akeneo/Component/FileStorage/Model/FileInterface.php (hunk 4 above)
Comment (jjanvier) [inline src/Akeneo/Component/FileStorage/Model/FileInterface.php]: yes it it for now
but I really hope to be able to drop it soon

I add a TODO

File: src/Akeneo/Component/FileStorage/PathGenerator.php (hunk 16 above)
Comment (jjanvier) [inline src/Akeneo/Component/FileStorage/PathGenerator.php]: ô_O the goal is to replace everything except `A-Za-z0-9.`

**Comment by jjanvier** [comment]
@jmleroux done

File: src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    /** @var MountManager */
+    protected $mountManager;
+
+    /** @var FileFactoryInterface */
+    protected $factory;
+
+    /**
+     * @param PathGeneratorInterface $pathGenerator
+     * @param MountManager           $mountManager
+     * @param SaverInterface         $saver
```
Comment (solivier) [inline src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php]: missing phpdoc

File: src/Akeneo/Bundle/FileStorageBundle/Doctrine/ORM/Repository/FileRepository.php [hunk 18]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Akeneo\Bundle\FileStorageBundle\Doctrine\ORM\Repository;
+
+use Akeneo\Component\FileStorage\Repository\FileRepositoryInterface;
+use Doctrine\ORM\EntityRepository;
+
+/**
+ * Implementation of ReferenceRepositoryInterface.
```
Comment (skeleton) [inline src/Akeneo/Bundle/FileStorageBundle/Doctrine/ORM/Repository/FileRepository.php]: nope

File: src/Akeneo/Bundle/FileStorageBundle/Resources/config/file_storage.yml [hunk 19]
```diff
@@ -0,0 +1,32 @@
+parameters:
+    akeneo_file_storage.file_storage.raw_file.storer.class:     Akeneo\Component\FileStorage\RawFile\RawFileStorer
+    akeneo_file_storage.file_storage.raw_file.fetcher.class:    Akeneo\Component\FileStorage\RawFile\RawFileFetcher
+    akeneo_file_storage.file_storage.file_factory.class:        Akeneo\Component\FileStorage\FileFactory
+    akeneo_file_storage.file_storage.file_mover.class:          Akeneo\Component\FileStorage\FileMover
+    akeneo_file_storage.file_storage.path_generator.class:      Akeneo\Component\FileStorage\PathGenerator
```
Comment (skeleton) [inline src/Akeneo/Bundle/FileStorageBundle/Resources/config/file_storage.yml]: extra spaces

File: src/Akeneo/Component/FileStorage/PathGenerator.php [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        $originalFileName = ($file instanceof UploadedFile) ? $file->getClientOriginalName() : $file->getFilename();
+        $guid = $this->generateId($originalFileName);
+        $sanitized = preg_replace('#[^A-Za-z0-9\.]#', '_', $originalFileName);
+
+        if (strlen($sanitized) > 100) {
+            $sanitized = sprintf('%s.%s', substr($sanitized, 0, 95), $file->getExtension());
+        }
+
+        $fileName = $guid.'_'.$sanitized;
```
Comment (skeleton) [inline src/Akeneo/Component/FileStorage/PathGenerator.php]: add spaces around "."

File: src/Akeneo/Component/FileStorage/PathGenerator.php [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+        $path = sprintf('%s/%s/%s/%s/', $guid[0], $guid[1], $guid[2], $guid[3]);
+        $pathName = $path.$fileName;
+
+        return [
+            'guid' => $guid,
+            'file_name' => $fileName,
+            'path' => $path,
+            'path_name' => $pathName,
```
Comment (skeleton) [inline src/Akeneo/Component/FileStorage/PathGenerator.php]: please align

File: src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+    public function __construct(
+        PathGeneratorInterface $pathGenerator,
+        MountManager $mountManager,
+        SaverInterface $saver,
+        FileFactoryInterface $factory
+    ) {
+        $this->pathGenerator = $pathGenerator;
+        $this->mountManager = $mountManager;
+        $this->saver = $saver;
+        $this->factory = $factory;
```
Comment (skeleton) [inline src/Akeneo/Component/FileStorage/RawFile/RawFileStorer.php]: please align

**Pull request merged**